### PR TITLE
Add PR 518 camera model support to Gaussian splat reconstruction

### LINKED ISSUE #252: Support world-space Gaussian splat reconstruction with unified camera model
# World-Space Gaussian Splat Reconstruction

Reality-capture should support world-space Gaussian rasterization, distortion-aware cameras, and MCMC on top of the merged `fvdb-core` camera/render API.

The current `SfmScene` API assumes undistorted pinhole cameras, and the repo is stll written against the older `fvdb-core` API based only on pixel-space rasterization.

The proposed solution is a moderate refactor:

- Similar to how optimizer objects are composed into `GaussianSplatReconstruction`, allow composition of a `RenderBackend` to select between world and camera/pixel space rendering. 
- Update `SfmCameraMetadata` to store an instance of the new`fvdb.CameraModel` to support multiple camera types in the `SfmScene`
- Add a scene transform to apply pixel space undistortion via `cv2.remap` rather than doing this in the dataset as is done today. This lets the user use pinhole cameras on undistorted images if they want rather than it being hardcoded in the dataset.
- Allow the user to select which render backend to use in training and which projection method each render backend can use (UT-Projection/World-Rasterization, Classic-Projection/World-Rasterization, Classic-Projection/Pixel-Rasterization). Note that not all combinations are valid, and not all combinations work with all camera types).

This work has two outputs:

 1. Changes to frgs API which allow a user to select rasterization and projection approach
 2. Public API changes in `GaussianSplatReconstruction` to allow user to select rasterization and projection approach
 3. A repo-wide migration of direct Gaussian render/projection call sites onto the new public `fvdb-core` API (PR 518)
 4. Addition of an unprojection `SfmScene` transform that converts images with distortion into their equivalent pinhole projection images 

## Assumptions

This design assumes the changes in`fvdb-core` PR 518

### PR DESCRIPTION
**Pull request by fwilliams**
## Summary
This PR addresses [#252](https://github.com/openvdb/fvdb-reality-capture/issues/252): support world-space Gaussian splat reconstruction on top of the new `fvdb-core` camera/render API, while cleaning up how reality-capture represents cameras, distortion, and render-time camera behavior.

At a high level, the existing Gaussian splat pipeline was too narrow for the `fvdb-core#518` API surface. Reconstruction and downstream tools still assumed undistorted pinhole cameras, image-space rendering was effectively hard-coded, and pose-optimization / checkpoint-loading behavior was not robust enough once the new camera/render paths were exercised more seriously.

This change keeps `GaussianSplatReconstruction` as the top-level coordinator, but introduces a narrow internal render seam plus a cleaner camera contract so the pipeline can support both image-space and world-space rendering while remaining readable and testable.

## What changed
- Added an internal render backend abstraction for Gaussian splat reconstruction, with image-space and world-space implementations.
- Updated reconstruction and evaluation paths to route rendering through that backend seam rather than hard-coding one rendering path.
- Simplified `SfmCameraMetadata` so it represents a single canonical pixel space and uses canonical `fvdb.CameraModel` plus packed `distortion_coeffs`.
- Removed the dataset-level undistortion flag and instead added an `UndistortImages` scene transform that materializes a new undistorted scene representation.
- Updated `SfmDataset` to operate directly on the current scene pixel space rather than switching between raw and derived camera states internally.
- Moved initialization parameters `initial_opacity` and `initial_covariance_scale` to optimizer config as the single source of truth, with backward-compatible checkpoint loading.
- Added explicit pinhole-camera validation for TSDF fusion and point-cloud extraction paths that still rely on pinhole depth unprojection assumptions.
- Updated benchmark config contracts and comparative benchmark utilities to match the optimizer-config move for initialization parameters.
- Added and updated unit tests for camera metadata, the undistortion transform, TSDF/point-cloud camera-model validation, optimizer helpers, and benchmark behavior.
- Improved render backend naming and documentation (`image`, `alpha`, `depth`) and cleaned up test helper typing.

## Follow-up fixes included in this branch
While validating the world-space and MCMC paths, this branch also picked up several correctness fixes that are worth calling out explicitly:
- Fixed pose optimization bookkeeping to use scene-global image IDs instead of training-local IDs.
- Sized the learned pose table against `sfm_scene.num_images` so pose indexing remains consistent across train / validation splits.
- Added a warning when pose optimization is enabled with a holdout validation set, since validation views still use their original poses.
- Added backward-compatible remapping for legacy checkpoints whose pose embeddings were stored using training-local IDs.
- Fixed `GaussianSplatReconstruction.from_state_dict()` so CPU-loaded checkpoints can be restored onto CUDA correctly by re-deviceing and re-leafing nested tensor state before optimizer reconstruction.

## Why this approach
The main goal of [#252](https://github.com/openvdb/fvdb-reality-capture/issues/252) was not just to "make the new API compile," but to make world-space Gaussian reconstruction a first-class path while keeping the pipeline understandable.

This PR takes a moderate-refactor approach:
- keep reconstruction orchestration centralized in `GaussianSplatReconstruction`
- introduce one narrow internal rendering seam instead of spreading conditional logic throughout training/eval
- make `fvdb.CameraModel` the canonical camera enum
- make undistortion a scene transform so `SfmScene`, `SfmCameraMetadata`, and `SfmDataset` all agree on a single pixel space at a time
- fail explicitly in downstream depth-unprojection paths that still only support pinhole cameras

That gives us world-space rendering support without turning reconstruction into a generic plugin framework, while also making future rendering backends easier to add.

## Notable bug fixes
- Fixed MCMC initialization so `initial_covariance_scale` and `initial_opacity` are taken from the optimizer config, instead of incorrectly falling back to reconstruction-config defaults.
- Fixed reconstruction training / metadata paths to use the correct optimized camera poses after pose optimization.
- Fixed checkpoint restore for legacy pose-optimization state and for CPU-loaded checkpoints restored onto CUDA.

## Camera/distortion behavior after this PR
- `SfmCameraMetadata` now uses canonical `fvdb.CameraModel` values and packed `distortion_coeffs` as the primary camera representation.
- If a workflow wants undistorted images, it should transform the scene with `UndistortImages` and then operate on that derived scene.
- If a workflow uses raw distorted cameras, the render backend is responsible for assembling the correct FVDB render arguments.
- TSDF/point-cloud extraction now reject unsupported non-pinhole camera models early instead of silently applying pinhole-only unprojection assumptions.

## Downstream impact
This PR also updates supporting tools and contracts so the new rendering/camera behavior is reflected consistently across the repo:
- CLI camera metadata loading
- meshing and point-cloud extraction validation
- benchmark schema/config extraction
- unit-test helpers and benchmark behavior when local benchmark artifacts are absent
- checkpoint compatibility for older reconstruction runs that predate scene-global pose IDs

## Blocked on
- `openvdb/fvdb-core#518`

## Issue
Closes #252.

## Test plan
- `source "/home/fwilliams/bin/miniconda3/etc/profile.d/conda.sh" && conda activate fvdb && python -m pytest -v`
- `source "/home/fwilliams/bin/miniconda3/etc/profile.d/conda.sh" && conda activate fvdb && frgs reconstruct-mcmc data/360_v2/garden -o scratch/exp_mcmc_init_paper_garden_no_pose_current_core_benchmark.ply -n exp_mcmc_init_paper_garden_no_pose_current_core_benchmark -vn 100 --cfg.max-epochs 60 --cfg.eval-at-percent 10 20 30 40 50 60 70 80 90 100 --cfg.save-at-percent 100 --cfg.batch-size 1 --cfg.render-backend world_space --cfg.projection-method unscented --cfg.no-optimize-camera-poses --opt.max-gaussians 1000000 --opt.initial-opacity 0.5 --opt.initial-covariance-scale 0.1 --io.log-path scratch/exp_mcmc_init_paper_garden_no_pose_current_core_benchmark`
- `source "/home/fwilliams/bin/miniconda3/etc/profile.d/conda.sh" && conda activate fvdb && pytest unit/test_gaussian_splat_3d.py -k projection_method_resolution_and_metadata -v` (run from `../fvdb-core/tests` after rebuilding/installing `fvdb-core` with `./build.sh install`)

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: fvdb_reality_capture/sfm_scene/sfm_metadata.py [hunk 1]
```diff
@@ -1,62 +1,111 @@
...
 
-    SIMPLE_PINHOLE = "SIMPLE_PINHOLE"
-    """
-    A simplified pinhole camera model with a single focal length and no lens distortion. The principal point is at the image center.
-    """
 
-    SIMPLE_RADIAL = "SIMPLE_RADIAL"
-    """
-    A simplified radial distortion camera model with a single focal length and one radial distortion coefficient.
+def _normalize_packed_distortion_coeffs(
```
Comment (swahtz) [inline fvdb_reality_capture/sfm_scene/sfm_metadata.py]: **💡 suggestion:** ‏ 'normalize' makes it sound like this function is normalizing the values of the coefficients to some unit length.  I'd suggest we change the name to `_as_packed_distortion_coeffs` or `_coerce_packed_distortion_coeffs` so the reader gets a better understanding of its functionality.

File: fvdb_reality_capture/sfm_scene/sfm_metadata.py [hunk 2]
```diff
@@ -300,71 +323,56 @@ def height(self) -> int:
...
+        return self._distortion_coeffs
 
     @property
-    def distortion_parameters(self) -> np.ndarray:
+    def can_undistort(self) -> bool:
         """
-        Return the distortion parameters of the camera.
-
-        The distortion parameters are used to correct lens distortion in the captured images.
+        Return whether this camera can be undistorted by the current OpenCV-based transform.
```
Comment (swahtz) [inline fvdb_reality_capture/sfm_scene/sfm_metadata.py]: **❓ question:** ‏ What is meant by the 'current' OpenCV-bsaed transform?  Is it the currently set values for the coefficients?

I also find it confusing this returns True if there is no distortion, surely then it can't undistort because there's no distortion?

File: fvdb_reality_capture/transforms/undistort_images.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+                assert image is not None, f"Failed to load image {image_meta.image_path}"
+                pbar.set_description(f"Undistorting {image_filename}")
+
+                undistorted_camera, undistort_map_x, undistort_map_y, undistort_roi = new_camera_metadata[
+                    image_meta.camera_id
+                ]
+                undistorted_image = self._undistort_and_crop(
+                    image, undistort_map_x, undistort_map_y, undistort_roi, self._remap_interpolation
+                )
+                cache_image_filename = f"image_{image_meta.image_id:0{num_zeropad}}"
```
Comment (swahtz) [inline fvdb_reality_capture/transforms/undistort_images.py]: **🚩 issue:** ‏ Here the image filename being written in the cache is based on the `image_meta.image_id` value but the cached filename checks were done on line 219 with the range for loop's iterator counter.  COLMAP scenes with 1-based or non-contiguous IDs are going to cause this to always regenerate the cache.

File: fvdb_reality_capture/radiance_fields/_gaussian_rendering.py [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+            distortion_coeffs = distortion_coeffs.contiguous()
+            height, width = datum["image"].shape[:2]
+            camera_model_enum = CameraModel(camera_model)
+            distortion_coeffs_arg = _distortion_coeffs_for_batch(camera_model_enum, distortion_coeffs)
+            render_function = (
+                model.render_images_and_depths_from_world
+                if config.sparse_depth_reg > 0.0
+                else model.render_images_from_world
+            )
+            render_function(
```
Comment (swahtz) [inline fvdb_reality_capture/radiance_fields/_gaussian_rendering.py]: **☑️ todo:** ‏ As above, we should wrap this in a `torch.no_grad` or perhaps decorate the whole function

File: tests/unit/test_training.py
Comment (swahtz) [inline tests/unit/test_training.py]: **☑️ todo:** ‏ There aren't any tests for the new `camera_model` `distortion_coeffs` fields on `SfmDataset`; it probably makes the most sense to add some small tests `test_training`.

**Review by swahtz — APPROVED** [review]
Overall looks great.  Just a few bugs I caught and some suggestions.

**Review by swahtz — APPROVED** [review]
I already approved, do I need to approve again?

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 5]
```diff
@@ -28,20 +28,18 @@ class GaussianSplatOptimizerMCMCConfig(GaussianSplatOptimizerConfig):
     Parameters for configuring the ``GaussianSplatOptimizerMCMC``.
     """
 
-    # Override base class defaults to match the MCMC paper:
```
Comment (harrism) [inline fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py:31]: **💡 suggestion:** ‏I wonder if this comment should be left in? Otherwise it might be hard to understand why MCMC overrides these values from the base class.

File: tests/benchmarks/contract.py [hunk 6]
```diff
@@ -21,7 +21,7 @@
     GaussianSplatReconstructionConfig,
 )
 
-CONTRACT_VERSION = 1
+CONTRACT_VERSION = 4
```
Comment (harrism) [inline tests/benchmarks/contract.py:24]: **❓ question:** ‏Why did this jump from 1 to 4?

File: tests/benchmarks/test_3dgs.py [hunk 7]
```diff
@@ -200,15 +204,31 @@ def create_benchmark_params():
             raise ValueError(f"No checkpoint paths specified for dataset: {dataset_name}")
 
         for checkpoint_path in checkpoint_paths:
-            params.append((dataset_path, run_path, checkpoint_path))
+            checkpoint_id = f"{pathlib.Path(dataset_path).name}-{pathlib.Path(checkpoint_path).parent.name}"
```
Comment (harrism) [inline tests/benchmarks/test_3dgs.py]: This change broke nightly benchmarks. Fixed in #278
